File: .mise/lib/imports/report.py
"""
Render one width's findings, the breaks grouped by the frame and rules they
share, each group with the hunk around the row it names.
"""

from collections     import defaultdict
from collections.abc import Callable
from difflib         import SequenceMatcher

from records         import Break, Width

CONTEXT = 3
SHOWN   = 30


def hunk(pairs: SequenceMatcher, row: int | None) -> list[str]:
    """
    Return the diff hunk of `pairs` holding row `row` of its second
    sequence, the nearest hunk where none holds it and the first where the
    row is unknown, as unified-diff lines cut to `CONTEXT` lines either side
    of the row, the rows between the hunk and `row` appended where the hunk
    ends short of it.
    """
    groups = list(pairs.get_grouped_opcodes(CONTEXT))
    if not groups:
        return []
    at = None if row is None else row - 1

    def distance(group: list[tuple]) -> int:
        first, last = group[0][3], group[-1][4]
        if at is None or first <= at < last:
            return 0
        return min(abs(at - first), abs(at - last))

    group          = min(groups, key=distance)
    i1, i2, j1, j2 = group[0][1], group[-1][2], group[0][3], group[-1][4]
    shown          = [(None, f"@@ -{i1 + 1},{i2 - i1} +{j1 + 1},{j2 - j1} @@")]
    for tag, a1, a2, b1, b2 in group:
        if tag in ("delete", "replace"):
            shown += [(None, f"-{line}") for line in pairs.a[a1:a2]]
        if tag != "delete":
            mark   = " " if tag == "equal" else "+"
            rows   = pairs.b[b1:b2]
            shown += [(b1 + k, f"{mark}{line}") for k, line in enumerate(rows)]
    if at is not None and j2 <= at:
        gap = range(max(j2, at - CONTEXT), at + 1)
        if gap.start > j2:
            shown.append((None, "..."))
        shown += [(k, f" {pairs.b[k]}") for k in gap]
    if at is not None and at < j1:
        gap  = range(max(0, at - CONTEXT), min(j1, at + CONTEXT + 1))
        lead = [(k, f" {pairs.b[k]}") for k in gap]
        if gap.stop < j1:
            lead.append((None, "..."))
        shown[1:1] = lead
    found     = (k for k, (seen, _) in enumerate(shown) if seen == at)
    index     = 1 + CONTEXT if at is None else next(found, 1 + CONTEXT)
    low, high = max(1, index - CONTEXT), index + CONTEXT + 1
    lines     = [line for _, line in shown]
    return [
        lines[0],
        *(["..."] if low > 1 else []),
        *lines[low:high],
        *(["..."] if high < len(lines) else [])
    ]


def render(reproduce: Callable[[str], str], found: Width, carried: set[str]) -> str:
    """
    Return the report of one width's findings, `carried` naming the breaks
    the baseline already holds, the breaks grouped by the frame and rules
    they share.
    """
    timing       = [brk for brk in found.breaks if brk.formatted.kind == "timeout"]
    raising      = [brk for brk in found.breaks if brk.formatted.kind != "timeout"]
    frames       = {(brk.frame, brk.attribution) for brk in found.breaks}
    uncomparable = "unmeasured" if found.unmeasured else found.uncomparable
    lines        = [
        f"  candidates   {found.candidates:>5}",
        f"  comparable   {found.comparable:>5}",
        f"  uncomparable {uncomparable:>5}",
        f"  breaks       {len(found.breaks):>5}",
        f"  frames       {len(frames):>5}",
        f"  timeouts     {len(timing):>5}",
        f"  flaky        {len(found.flaky):>5}"
    ]
    if carried:
        lines.append(f"  carried      {len(carried):>5}")
    for heading, listed in (("raises or rebinds", raising), ("times out", timing)):
        if not listed:
            continue
        groups = defaultdict(list)
        for brk in listed:
            groups[(brk.frame, brk.attribution)].append(brk)
        lines += ["", f"  {heading} ({len(listed)} modules at {len(groups)} frames):"]
        for (frame, attribution), members in groups.items():
            lines += rendered_group(
                attribution = attribution,
                carried     = carried,
                frame       = frame,
                members     = members,
                reproduce   = reproduce
            )
    if found.flaky:
        lines += ["", f"  flaky, a second run disagreed ({len(found.flaky)}):"]
        lines += [f"    {module}" for module in found.flaky]
    if found.unmeasured:
        lines += ["", f"  unmeasured, a run left no record ({len(found.unmeasured)}):"]
        lines += [f"    {module}" for module in found.unmeasured]
    return "\n".join(lines)


def rendered_group(
    reproduce   : Callable[[str], str],
    members     : list[Break],
    frame       : tuple[str, int | None],
    attribution : str,
    carried     : set[str]
) -> list[str]:
    """
    Return the report of the breaks `members` sharing `frame` and
    `attribution`, the hunk once, a reason every member shares once, the
    modules the baseline does not carry ahead of the ones it does, and up to
    `SHOWN` modules named with the rest counted.
    """
    file, row = frame
    where     = f"{file}:{row}" if row else file
    reasons   = {brk.reason for brk in members}
    shared    = members[0].reason if len(reasons) == 1 else ""
    members   = sorted(members, key=lambda brk: (brk.module in carried, brk.module))
    lines     = [f"    {where} {attribution}"]
    if shared:
        lines.append(f"      each {shared}")
    lines += [f"      {line}" for line in members[0].hunk]
    for brk in members[:SHOWN]:
        reason   = "" if shared else f" {brk.reason}"
        carriage = ", carried by the baseline" if brk.module in carried else ""
        lines.append(f"      {brk.module}{reason}{carriage}")
    if len(members) > SHOWN:
        lines.append(f"      ... and {len(members) - SHOWN} more")
    lines.append(f"      reproduce with {reproduce(members[0].module)}")
    return lines

File: crate/tests/imports/report.rs
//! Rendering one width's findings. The breaks are tallied by the frame and
//! rules they share, each shown with the hunk around the row it names and the
//! command that reproduces one of its modules alone. Beneath them the report
//! lists each file the pipeline could not format, each module whose original
//! run did not end cleanly, and each name the comparison left out.

use std::{
    collections::BTreeMap,
    fmt::{Display, Write},
    path::Path,
};

use itertools::Itertools;

use crate::{
    common::{Hit, SHOWN, Tally, WIDTHS_VAR, remainder, setting},
    corpus::PYTHON_VAR,
    execute::TIMEOUT_VAR,
    outcome::Kind,
    records::{Break, Frame, Width},
    sweep::DEFAULT_LABEL,
};

/// Renders one width's findings.
pub(crate) fn render(found: &Width) -> String {
    let (raising, rebinding, timeouts) = tallied(found);
    let uncomparable = if found.unmeasured.is_empty() {
        found.uncomparable.len().to_string()
    } else {
        "unmeasured".to_owned()
    };
    let row = |label: &str, value: &dyn Display| format!("  {label:<12} {value:>5}");
    let mut lines = vec![
        row("candidates", &found.candidates),
        row("rewritten", &found.rewritten),
        row("comparable", &found.comparable),
        row("uncomparable", &uncomparable),
        row("breaks", &found.broken()),
        row("raises", &found.counting(Kind::Raised)),
        row("rebinds", &found.counting(Kind::Ok)),
        row("timeouts", &found.counting(Kind::Timeout)),
        row("rejected", &found.rejected.len()),
        row("flaky", &found.flaky.len()),
        row("varying", &found.varying()),
        row("left out", &found.left_out()),
    ];
    if found.unread > 0 {
        lines.push(row("unread", &found.unread));
    }
    let mut rendered = lines.join("\n");
    if !found.uncomparable.is_empty() {
        let split = found
            .reaches()
            .iter()
            .map(|(reach, counted)| format!("{counted} {reach}"))
            .join(", ");
        let _ = write!(
            rendered,
            "\n\nuncomparable by reach ({}):\n  {split}",
            found.uncomparable.len(),
        );
    }
    rendered.push_str(&raising.render("raises"));
    rendered.push_str(&rebinding.render("runs and binds a different namespace"));
    rendered.push_str(&timeouts.render("times out"));
    let rejected: Vec<_> = found
        .rejected
        .iter()
        .map(|(file, error)| format!("{file}  {error}"))
        .collect();
    let blocked: Vec<_> = found
        .uncomparable
        .iter()
        .map(|(module, left)| format!("{module}  {}, {}", left.reach, left.reason))
        .collect();
    let left_out: Vec<_> = found
        .removed
        .values()
        .flat_map(BTreeMap::values)
        .cloned()
        .collect();
    let varied = excluded(found);
    for (heading, listed) in [
        ("rejected, the pipeline could not format it", &rejected),
        ("uncomparable, the original did not run cleanly", &blocked),
        ("left out, a recorded fix removed the binding", &left_out),
        ("flaky, a second run varied", &varied),
        ("unmeasured, a run left no record", &found.unmeasured),
    ] {
        if !listed.is_empty() {
            let _ = write!(
                rendered,
                "\n\n{heading} ({}):\n  {}{}",
                listed.len(),
                listed.iter().take(SHOWN).format("\n  "),
                remainder(listed.len()),
            );
        }
    }
    rendered
}

/// The sentence naming where a break raises, why it raises, and what it
/// traces to, which is the wording a tally keys it under.
fn defect(brk: &Break) -> String {
    let Frame { file, row } = &brk.frame;
    let at = row.map_or_else(|| file.clone(), |row| format!("{file}:{row}"));
    format!("{at} {}, {}", brk.reason, brk.attribution)
}

/// The rows naming each module a run set aside, one per module beside the
/// names it varied on. A module whose whole namespace varied carries no
/// names, so its row reads `the whole namespace` in their place.
fn excluded(found: &Width) -> Vec<String> {
    found
        .flaky
        .iter()
        .map(|(module, names)| {
            if names.is_empty() {
                format!("{module}  the whole namespace")
            } else {
                format!("{module}  {}", names.iter().format(", "))
            }
        })
        .collect()
}

/// The command that runs one module on its own, carrying every knob the
/// current run set.
fn reproduction(label: &str, module: &str) -> String {
    let mut knobs: Vec<_> = [PYTHON_VAR, TIMEOUT_VAR]
        .iter()
        .filter_map(|knob| Some(format!("{knob}={}", setting(knob)?)))
        .collect();
    if label != DEFAULT_LABEL {
        knobs.push(format!("{WIDTHS_VAR}={label}"));
    }
    knobs.push("mise run imports".to_owned());
    format!("{} {module}", knobs.join(" "))
}

/// The breaks, split by how the formatted run ended, each keyed by the
/// sentence they share so one frame reaching many modules reports once.
fn tallied(found: &Width) -> (Tally, Tally, Tally) {
    let mut raising = Tally::default();
    let mut rebinding = Tally::default();
    let mut timeouts = Tally::default();
    for brk in &found.breaks {
        let tally = match brk.formatted.kind {
            Kind::Ok => &mut rebinding,
            Kind::Raised => &mut raising,
            Kind::Timeout => &mut timeouts,
            Kind::Unmeasured => {
                unreachable!(
                    "invariant: a run that left no record is unmeasured rather than broken"
                )
            }
        };
        tally.record_hit(
            defect(brk),
            Path::new(&brk.module),
            Hit {
                clause: None,
                detail: Some(brk.hunk.join("\n")),
                repro: Some(reproduction(&found.label, &brk.module)),
            },
        );
    }
    (raising, rebinding, timeouts)
}
